### PR TITLE
feat(metrics): update pro engine config metrics

### PR DESCRIPTION
**Pull request by kopecs**
Updates semgrep to send the updated metrics about engine configuration added to interfaces. This only affects pro scans since no additional information is present in the oss config.

- [x] PRIVACY.md updates.

### CODE REVIEW

**Comment by kopecs** [comment]
Not sure what the status of metrics testing is. In any event, this only substantively effects pro, so tests would end up there. Sounds like @emjin was doing some work there?

Edit: Seems like we get some oss coverage based on snapshot.

**Comment by amchiclet** [comment]
> Not sure what the status of metrics testing is

Very good question that I'd also be interested in knowing the answer. If there was a way to dump the metrics to a local file or stdout, that would be a good start. You could just paste the output as a comment for now. Automating this, or whether we need automation, is also unknown for me. Sorry if this answer wasn't helpful.

**Comment by kopecs** [comment]
I had to update a snapshot, so we at least have some. I would guess we could improve it quite a bit though.

File: cli/src/semgrep/run_scan.py
Comment (amchiclet) [inline cli/src/semgrep/run_scan.py]: Optional: nested conditional expressions, especially the python way, make my brain explode,.But this specific occurrence is manageable. Others might also be ok, so feel free to ignore and resolve.

File: changelog.d/ea-297.added
Comment (amchiclet) [inline changelog.d/ea-297.added]: nit: valdiation -> validation

Also, can you maybe add one more sentence why tracking this extra information is helpful to the user?

File: scripts/lint-ocaml
Comment (amchiclet) [inline scripts/lint-ocaml]: :)

Just smiling. You can resolve this.

**Review by amchiclet — APPROVED** [review]
I think it's good to go. If you don't mind waiting until @emjin reviews it, or if not until tomorrow PST to merge it, I think we're fine.

**Comment by amchiclet** [comment]
Also not sure why `pre-commit` is failing. If that can be fixed or an explanation added why it's not related to this PR, that would help.

File: scripts/lint-ocaml
Comment (kopecs) [reply scripts/lint-ocaml]: Ah I didn't even mean to commit this. I just keep having trouble getting bash 5.x to be reliably used instead of the default. 😔

**Comment by kopecs** [comment]
Yeah I'm not sure. Will investigate tomorrow.

File: cli/src/semgrep/run_scan.py
Comment (kopecs) [reply cli/src/semgrep/run_scan.py]: This is definitely close to the limit, imo, but I think this is more understandable than introducing an if statement into this block.

File: cli/src/semgrep/metrics.py
Comment (emjin) [inline cli/src/semgrep/metrics.py:175]: I'm a little confused. Don't we create classes for the atd types? Can we use that class instead?

File: cli/src/semgrep/metrics.py
Comment (emjin) [inline cli/src/semgrep/metrics.py]: At minimum, needs an explanation of what `secrets_origins` is. Ideally this is structured in a way that makes that easier to discover.

File: cli/src/semgrep/run_scan.py
Comment (emjin) [inline cli/src/semgrep/run_scan.py]: To be honest, if I didn't have some secrets context, I think after reading this multiple times I still would not understand how to interpret this wrt answering the question "Did secrets run?". Would like to simplify the type

**Review by emjin — CHANGES_REQUESTED** [review]
Mostly an extension of comments in the semgrep-interface PR, but I wanted to point out the places where I felt confused in this one

File: cli/src/semgrep/metrics.py
Comment (kopecs) [reply cli/src/semgrep/metrics.py:175]: Yes, we do, but this class is used instead of the atd generated thing. I could try and migrate everything to use the metrics type instead but this has a couple issues:
- it makes business logic depend on the definition as it exists in the atd
- it makes more difficult to have methods on the relevant type
- it requires more changes to use the atd based class instead of the currently used one

File: cli/src/semgrep/metrics.py
Comment (kopecs) [reply cli/src/semgrep/metrics.py]: Commenting on the other one; can add a docstring here.

File: cli/src/semgrep/run_scan.py
Comment (kopecs) [reply cli/src/semgrep/run_scan.py]: I am not sure exactly what simplification you would like: ultimately a `None` for `secrets_config` would mean secrets was not run. `Some { secrets_origins = "..." }` would correspond to secrets running with that option for the origins. I think what the origins mean is not obvious, but that seems a bit distinct.

Comment (emjin) [reply cli/src/semgrep/run_scan.py]: I think the nicest thing would be to have a class for the secrets config type which is also passed to the semgrep-pro engine, and which implements `str`. Don't know how practical this is given the existing architecture; would hope that it isn't too hard.

**Review by emjin — APPROVED** [review]
Looks good for a first pass. I'd still like stronger guarantees in some form that the metrics will reflect what secrets sends. e.g. let's say for some reason we decided to add an environment variable that also can disable secrets validation, I'd want it to be easy to figure out that we need to update the metrics.

File: cli/src/semgrep/run_scan.py
Comment (kopecs) [reply cli/src/semgrep/run_scan.py]: Just changed to an extensible sum.

**Comment by kopecs** [comment]
Merging with CI failing:
 - pro test & pro rules: will be fixed by https://github.com/semgrep/semgrep-proprietary/pull/1191
 - ~~pre-commit: see https://github.com/semgrep/semgrep/pull/9495/commits/be4f768cc56a8790acc6269d24db8ad59fee380d~~ rebasing to get the change.

**Pull request merged**
